src/main/utility/synchronization/simple_latch.rs

```diff
@@ -106,27 +109,41 @@ impl LatchWaiter {
                 _ => panic!("Latch has been opened multiple times without us waiting"),
             }
 
-            let futex_word: &AtomicU32 = self.latch_gen.as_ref();
-
-            let rv = Errno::result(unsafe {
-                libc::syscall(
-                    libc::SYS_futex,
-                    futex_word as *const AtomicU32 as *const u32,
-                    libc::FUTEX_WAIT | libc::FUTEX_PRIVATE_FLAG,
-                    latch_gen,
-                    std::ptr::null() as *const libc::timespec,
-                    std::ptr::null_mut() as *mut u32,
-                    0u32,
-                )
-            });
-            assert!(
-                rv.is_ok() || rv == Err(Errno::EAGAIN) || rv == Err(Errno::EINTR),
-                "FUTEX_WAIT failed with {rv:?}"
-            );
+            if !self.spin_yield {
+                let futex_word: &AtomicU32 = self.latch_gen.as_ref();
+
+                let rv = Errno::result(unsafe {
+                    libc::syscall(
+                        libc::SYS_futex,
+                        futex_word as *const AtomicU32 as *const u32,
+                        libc::FUTEX_WAIT | libc::FUTEX_PRIVATE_FLAG,
+                        latch_gen,
+                        std::ptr::null() as *const libc::timespec,
+                        std::ptr::null_mut() as *mut u32,
+                        0u32,
+                    )
+                });
+                assert!(
+                    rv.is_ok() || rv == Err(Errno::EAGAIN) || rv == Err(Errno::EINTR),
+                    "FUTEX_WAIT failed with {rv:?}"
+                );
+            } else {
+                // we don't know if a pause instruction is beneficial or not here, but it doesn't
+                // seem to hurt performance
+                // https://www.intel.com/content/www/us/en/docs/cpp-compiler/developer-guide-reference/2021-9/pause-intrinsic.html
+                std::hint::spin_loop();
+                std::thread::yield_now();
+            }
         }
 
         self.gen = self.gen.wrapping_add(1);
     }
+
+    /// When waiting, should we `sched_yield` in a spinloop indefinitely rather than futex-wait?
+    /// This may improve performance in some workloads. This is disabled by default.
+    pub fn enable_spinning(&mut self, value: bool) {
```

Comment:
Nit - is there a reason to have this as a method instead of a constructor parameter? I don't think we'll ever want to change it for a latch once constructed, will we?

Reply:
My reasoning was:

- The waiter should choose whether it spins or not, instead of the latch. A `LatchWaiter` can only be constructed through `Latch::waiter(&mut self)`, so having an `enable_spinning` method on the waiter allows you to make the decision closer to where you'll actually be doing the waiting.
- Enabling spinning is an optional feature, so I think it's nicer to have this feature be opt-in with a reasonable default (disabled), rather than requiring the constructor to specify explicitly if it's enabled or disabled (setting it in the constructor doesn't allow for a default behaviour, unless maybe we made it an `Option<bool>`).

Reply:
> so having an enable_spinning method on the waiter allows you to make the decision closer to where you'll actually be doing the waiting.

Oops, I think what I meant was to add the field to the `waiter` method, not to the latch constructor.

> Enabling spinning is an optional feature, so I think it's nicer to have this feature be opt-in with a reasonable default (disabled),

Hmm, another way to achieve this would be for `waiter` to return a non-spinning waiter, and for a new method like `spinning_waiter` to return one with spinning enabled.

I'm also fine with the code as is, though.